### PR TITLE
 add test for multidatasample

### PR DESCRIPTION
**Pull request by marouaneamz**
#5

### CODE REVIEW

**Comment by marouaneamz** [comment]
thank you for reviewing the test

File: tests/test_structures/test_datasample.py [hunk 1]
```diff
@@ -156,3 +157,20 @@ def test_set_pred_score(self):
         data_sample = ClsDataSample()
         data_sample.set_pred_score(torch.tensor([0.1, 0.1, 0.6, 0.1, 0.1]))
         self.assertEqual(data_sample.pred_label.num_classes, 5)
+
+
+class TestMultiTaskDataSample(GenericTest):
+    def test_set_gt_label(self):
+        self._test_set_label(key='gt_label',data_sample=MultiTaskDataSample().to_cls_data_samples())
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]: On my current understanding, here you are testing that you can set 
* a number
* a tensor number
* an array ....

like `multi_task_data_sample.set_get_label(number)`

I feel this is not what we want to achieve, please rework the whole unit test strategy

File: tests/test_structures/test_datasample.py [hunk 2]
```diff
@@ -156,3 +157,20 @@ def test_set_pred_score(self):
...
+        self._test_set_label(key='pred_label',data_sample=MultiTaskDataSample().to_cls_data_samples())
+
+    def test_get_task_mask(self):
+        data_samples = []
+        gt_label = {}
+        gt_label['task1'] = 1
+        data_sample = MultiTaskDataSample(['task1']).to_cls_data_samples().set_gt_label(gt_label)
+        data_samples.append(data_sample)
+        self.assertTrue(data_samples.get_task_mask('task1'),True)
+        self.assertFalse(data_samples.get_task_mask('task2'),False)
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]: Let's test first without any tasks as input like `MultiTaskDataSample()`

we could add the tasks optionnaly `MultiTaskDataSample(metainfo={
  'task0': {num_classes='10'},
  'task1': {}
})`


```suggestion
        data_sample = MultiTaskDataSample().set_gt_label(gt_label)
        self.assertTrue(data_sample.get_task_mask('task1'),True)
        self.assertTrue(data_sample.get_task_mask('task2'),False)
```

File: tests/test_structures/test_datasample.py [hunk 3]
```diff
@@ -156,3 +157,20 @@ def test_set_pred_score(self):
         data_sample = ClsDataSample()
         data_sample.set_pred_score(torch.tensor([0.1, 0.1, 0.6, 0.1, 0.1]))
         self.assertEqual(data_sample.pred_label.num_classes, 5)
+
+
+class TestMultiTaskDataSample(GenericTest):
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]: please add the following tests : 

### to_target_data_sample

* `data_sample.to_target_data_sample('ClsDataSample', 'task0')`
  * if gt_label['task0'] is a number with 'ClsDataSample' -> return a ClsDataSample
  * if gt_label['task0'] is a number bigger than 'num_data_sample' with 'ClsDataSample'  with MultiTaskDataSample(metainfo={ 'task0': {num_classes='10'}, 'task1': {} })-> throw an error
  * if gt_label['task0'] is a string -> throw an error
  * if 'task0' is not in gt_label -> throw an error
 
* `data_sample.to_target_data_sample('MultiTaskDataSample', 'nested_task')` -> return an other  'MultiTaskDataSample' with `get_gt_label` as expected
* `data_sample.to_target_data_sample('MultiTaskDataSample', 'nested_task')` throw an error if the gt_label for key `nested_task` is a number (and not a dictonnary)

File: tests/test_structures/test_datasample.py (hunk 1 above)
Comment (piercus) [reply tests/test_structures/test_datasample.py]: you need to reimplement `_test_set_label` for `MultiTaskDataSample` and test the following polymorphism behavior : 

| Label set Input | metainfo | Expected get output |
|---|---|---|
| Dict | None | LabelData |
| Empty Dict | None | Empty LabelData |
| Dict with proper keys | metainfo | LabelData |
| Empty Dict | metainfo | Empty LabelData |
| Dict with outside keys | metainfo | Throw an error |

**Review by piercus — CHANGES_REQUESTED** [review]

File: tests/test_structures/test_datasample.py [hunk 4]
```diff
@@ -5,12 +5,12 @@
...
-from mmcls.structures import ClsDataSample
+from mmcls.structures import ClsDataSample, MultiTaskDataSample
 
 
-class TestClsDataSample(TestCase):
 
-    def _test_set_label(self, key):
+class TestClsDataSample(TestCase):
+    def _test_set_label(self,key):
         data_sample = ClsDataSample()
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]: please checkout this part 

see https://stackoverflow.com/a/6963150

you should be able to do something like

```
git checkout -p my_multi_task_1_x-- tests/test_structures/test_datasample.py
```

File: tests/test_structures/test_datasample.py [hunk 5]
```diff
@@ -156,3 +156,89 @@ def test_set_pred_score(self):
...
+        self.assertIn(key, data_sample)
+        label = getattr(data_sample, key)
+        self.assertIsInstance(label, LabelData)
+
+        # Test empty Dict without metainfo
+        method({})
+        self.assertIn(key, data_sample)
+        label = getattr(data_sample, key)
+        self.assertIsInstance(label, LabelData)
+
```
Comment (piercus) [inline tests/test_structures/test_datasample.py:190]: move `data_sample2` init here

File: tests/test_structures/test_datasample.py [hunk 6]
```diff
@@ -156,3 +156,89 @@ def test_set_pred_score(self):
...
+    def _test_set_label(self,key):
+        data_sample = MultiTaskDataSample()
+        data_sample2 = MultiTaskDataSample(metainfo={ 'task0': {'num_classes' :10}, 'task1': {'num_classes':3} })
+        method = getattr(data_sample, 'set_' + key)
+        # Test Dict without metainfo
+        method({'task0': 0,'task1':2})
+        self.assertIn(key, data_sample)
+        label = getattr(data_sample, key)
+        self.assertIsInstance(label, LabelData)
```
Comment (piercus) [inline tests/test_structures/test_datasample.py:170]: please test `label.getattr('task0') == 1`

File: tests/test_structures/test_datasample.py [hunk 7]
```diff
@@ -156,3 +156,89 @@ def test_set_pred_score(self):
...
+
+        # Test empty Dict without metainfo
+        method({})
+        self.assertIn(key, data_sample)
+        label = getattr(data_sample, key)
+        self.assertIsInstance(label, LabelData)
```
Comment (piercus) [inline tests/test_structures/test_datasample.py:177]: please test `label.getattr('task0')` is raising an error

File: tests/test_structures/test_datasample.py [hunk 8]
```diff
@@ -156,3 +156,89 @@ def test_set_pred_score(self):
...
+
+        # Test Dict with metainfo
+        method({'task0': 0,'task1':2})
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]: please use method2

File: tests/test_structures/test_datasample.py [hunk 9]
```diff
@@ -156,3 +156,89 @@ def test_set_pred_score(self):
...
+        self.assertIn(key, data_sample2)
+        label = getattr(data_sample2, key)
+        self.assertIsInstance(label, LabelData)
+
+        # Test empty Dict with metainfo
+        method({})
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]: method2

File: tests/test_structures/test_datasample.py [hunk 10]
```diff
@@ -156,3 +156,89 @@ def test_set_pred_score(self):
...
+        self.assertIn(key, data_sample2)
+        label = getattr(data_sample2, key)
+        self.assertIsInstance(label, LabelData)
+
+        # Test Dict with metainfo
+        method({'task0': 0,'task3':2})
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]: method2

Comment (piercus) [reply tests/test_structures/test_datasample.py]: method2 should fail here

File: tests/test_structures/test_datasample.py [hunk 11]
```diff
@@ -156,3 +156,89 @@ def test_set_pred_score(self):
...
+        method({'task0': 0,'task1':2})
+        self.assertIn(key, data_sample2)
+        label = getattr(data_sample2, key)
+        self.assertIsInstance(label, LabelData)
+
+        # Test empty Dict with metainfo
+        method({})
+        self.assertIn(key, data_sample2)
+        label = getattr(data_sample2, key)
+        self.assertIsInstance(label, LabelData)
```
Comment (piercus) [inline tests/test_structures/test_datasample.py:201]: Please test that `label.getattr('task0')` is not raising an error

File: tests/test_structures/test_datasample.py [hunk 12]
```diff
@@ -156,3 +156,89 @@ def test_set_pred_score(self):
...
+        self._test_set_label(key='pred_label')
+
+    def test_get_task_mask(self):
+        gt_label = {}
+        gt_label['task0'] = 1
+        data_sample = MultiTaskDataSample().set_gt_label(gt_label)
+        self.assertTrue(data_samples.get_task_mask('task0'),True)
+        self.assertTrue(data_samples.get_task_mask('task1'),False)
+
+    def test_to_target_data_sample(self):
```
Comment (piercus) [inline tests/test_structures/test_datasample.py:227]: Good for me :-)

**Review by piercus — CHANGES_REQUESTED** [review]

**Comment by marouaneamz** [comment]
neglect the modifications on evaluation , it's just modifications to fix linter (I still can't pass the lint of mmlcs)

File: mmcls/structures/multi_task_data_sample.py [hunk 13]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
         return self
 
-    def set_pred_label(self, value: Dict) -> 'MultiTaskDataSample':
-        """Set label of ``pred_label``."""
-        if 'pred_label' in self:
-            self.pred_label.score = value
+    """
+    def set_pred_task(self, value: Dict) -> 'MultiTaskDataSample':
+        if 'pred_task' in self:
+            self.pred_task.score = value
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]:
```suggestion
            self.pred_task = value
```

File: mmcls/structures/multi_task_data_sample.py [hunk 14]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
         else:
-            self.pred_label = LabelData(score=value)
+            self.pred_task = LabelData(score=value)
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]:
```suggestion
            self.pred_task = LabelData(**value)
```

File: mmcls/structures/multi_task_data_sample.py [hunk 15]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
         return self
-
+    """
     def get_task_mask(self, task_name):
-        return task_name in self.gt_label
+        return task_name in self.gt_task.label
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]:
```suggestion
        return task_name in self.gt_task
```

File: mmcls/structures/multi_task_data_sample.py [hunk 16]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
 
+    """
     def get_task_sample(self, task_name):
-        label = getattr(self.gt_label, task_name)
-        label_task = ClsDataSample().set_gt_label(label)
+        label = self.gt_task.label[task_name]
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]:
```suggestion
        label = self.gt_task[task_name]
```

File: mmcls/structures/multi_task_data_sample.py [hunk 17]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
+    @pred_task.setter
+    def pred_task(self, value: LabelData):
+        self.set_field(value, '_pred_task', dtype=LabelData)
+
+    @pred_task.deleter
+    def pred_task(self):
+        del self._pred_task
+    """
+    def to_cls_data_samples(self, task_name):
+        label = self.gt_task.label[task_name]
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]:
```suggestion
        label = self.gt_task[task_name]
```

File: mmcls/structures/multi_task_data_sample.py [hunk 18]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
+        label_task = ClsDataSample(
+            metainfo=self.metainfo.get(task_name, {})).set_gt_label(
+                value=label)
+        return label_task
+
+    def to_multi_task_data_sample(self, task_name):
+        label = self.gt_task.label[task_name]
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]:
```suggestion
        label = self.gt_task[task_name]
```

File: mmcls/structures/multi_task_data_sample.py [hunk 19]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
+    @pred_task.deleter
+    def pred_task(self):
+        del self._pred_task
+    """
+    def to_cls_data_samples(self, task_name):
+        label = self.gt_task.label[task_name]
+        label_task = ClsDataSample(
+            metainfo=self.metainfo.get(task_name, {})).set_gt_label(
+                value=label)
+        return label_task
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]: please set also values from `self.pred_task[task_name]` into the output data_sample

File: mmcls/structures/multi_task_data_sample.py [hunk 20]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
+
+    def to_multi_task_data_sample(self, task_name):
+        label = self.gt_task.label[task_name]
+        label_task = MultiTaskDataSample().set_gt_task(value=label)
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]: please set also values from `self.pred_task[task_name]` into the output data_sample

File: tests/test_structures/test_datasample.py [hunk 21]
```diff
@@ -156,3 +156,106 @@ def test_set_pred_score(self):
...
+
+    def _test_set_label(self, key):
+        data_sample = MultiTaskDataSample()
+        method = getattr(data_sample, 'set_' + key)
+        # Test Dict without metainfo
+        method({'task0': 0, 'task1': 2})
+        self.assertIn(key, data_sample)
+        label = getattr(data_sample, key)
+        self.assertIsInstance(label, LabelData)
+        self.assertEqual(label.label['task0'], 0)
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]:
```suggestion
        self.assertEqual(label['task0'], 0)
```

File: tests/test_structures/test_datasample.py [hunk 22]
```diff
@@ -156,3 +156,106 @@ def test_set_pred_score(self):
...
+
+        # Test empty Dict without metainfo
+        method({})
+        self.assertIn(key, data_sample)
+        label = getattr(data_sample, key)
+        self.assertIsInstance(label, LabelData)
+        with self.assertRaises(Exception):
+            label.label['task0']
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]:
```suggestion
            label['task0']
```

File: tests/test_structures/test_datasample.py [hunk 23]
```diff
@@ -156,3 +156,106 @@ def test_set_pred_score(self):
...
+        label = getattr(data_sample2, key)
+        self.assertIsInstance(label, LabelData)
+
+        # Test empty Dict with metainfo
+        method2({})
+        self.assertIn(key, data_sample2)
+        label = getattr(data_sample2, key)
+        self.assertIsInstance(label, LabelData)
+        with self.assertRaises(Exception):
+            label.label['task0']
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]:
```suggestion
            label['task0']
```

File: mmcls/structures/multi_task_data_sample.py [hunk 24]
```diff
@@ -26,64 +21,83 @@ def format_task_label(value: Dict, tasks: List[str]) -> LabelData:
...
-    def set_gt_label(self, value: Dict) -> 'MultiTaskDataSample':
-        """Set label of ``gt_label``."""
-        label = format_task_label(value, self.tasks)
-        if 'gt_label' in self:
-            self.gt_label.label = label.label
+    def set_gt_task(self, value: Dict) -> 'MultiTaskDataSample':
+        """Set label of ``gt_task``."""
+        label = format_task_label(value, self.metainfo)
+        if 'gt_task' in self:
+            self.gt_task.label = label.label
```
Comment (piercus) [inline mmcls/structures/multi_task_data_sample.py]:
```suggestion
            self.gt_task = label
```

File: tests/test_structures/test_datasample.py [hunk 25]
```diff
@@ -156,3 +156,106 @@ def test_set_pred_score(self):
...
+
+    """
+    def test_set_pred_score(self):
+        data_sample = MultiTaskDataSample()
+        data_sample.set_pred_score(torch.tensor([0.1, 0.1, 0.6, 0.1, 0.1]))
+        self.assertIn('score', data_sample.pred_label)
+        torch.testing.assert_allclose(data_sample.pred_label.score,
+                                      [0.1, 0.1, 0.6, 0.1, 0.1])
+
+    """
```
Comment (piercus) [inline tests/test_structures/test_datasample.py]:
```suggestion
```

**Review by piercus — CHANGES_REQUESTED** [review]

**Review by piercus — APPROVED** [review]
